pkg/pipelines/scm/gitlab.go

```diff
@@ -0,0 +1,122 @@
+
+}
+
+// CreatePRBinding returns a TriggerBinding for GitLab merge request hooks
+func (repo *GitLabRepository) CreatePRBinding(ns string) (triggersv1.TriggerBinding, string) {
+	return triggersv1.TriggerBinding{
+		TypeMeta:   triggerBindingTypeMeta,
+		ObjectMeta: meta.ObjectMeta(meta.NamespacedName(ns, gitlabPRBindingName)),
+		Spec: triggersv1.TriggerBindingSpec{
+			Params: []triggersv1.Param{
+				createBindingParam("gitref", "$(body.object_attributes.source_branch)"),
+				createBindingParam("gitsha", "$(body.object_attributes.last_commit.id)"),
+				createBindingParam("gitrepositoryurl", "$(body.project.git_http_url)"),
+				createBindingParam("fullname", "$(body.project.path_with_namespace)"),
+			},
+		},
+	}, gitlabPRBindingName
+
+}
+
+// CreatePushBinding returns a TriggerBinding for GitLab push hooks
+func (repo *GitLabRepository) CreatePushBinding(ns string) (triggersv1.TriggerBinding, string) {
+	return triggersv1.TriggerBinding{
+		TypeMeta:   triggerBindingTypeMeta,
+		ObjectMeta: meta.ObjectMeta(meta.NamespacedName(ns, gitlabPushBindingName)),
+		Spec: triggersv1.TriggerBindingSpec{
+			Params: []triggersv1.Param{
+				createBindingParam("gitref", "$(body.ref)"),
+				createBindingParam("gitsha", "$(body.after)"),
+				createBindingParam("gitrepositoryurl", "$(body.project.git_http_url)"),
+			},
+		},
+	}, gitlabPushBindingName
+}
+
+// URL returns the URL of the GitLab repository
+func (repo *GitLabRepository) URL() string {
+	return repo.url.String()
+}
+
+// CreateCITrigger creates a CI eventlistener trigger for GitLab
+func (repo *GitLabRepository) CreateCITrigger(name, secretName, secretNS, template string, bindings []string) v1alpha1.EventListenerTrigger {
+	return triggersv1.EventListenerTrigger{
+		Name: name,
+		Interceptors: []*triggersv1.EventInterceptor{
+			createEventInterceptor(gitlabCIDryRunFilters, repo.path),
+			repo.CreateInterceptor(secretName, secretNS),
+		},
+		Bindings: createBindings(bindings),
+		Template: createListenerTemplate(template),
+	}
+}
+
+// CreateCDTrigger creates a CD eventlistener trigger for GitLab
+func (repo *GitLabRepository) CreateCDTrigger(name, secretName, secretNS, template string, bindings []string) v1alpha1.EventListenerTrigger {
+	return triggersv1.EventListenerTrigger{
+		Name: name,
+		Interceptors: []*triggersv1.EventInterceptor{
+			createEventInterceptor(gitlabCDDeployFilters, repo.path),
+			repo.CreateInterceptor(secretName, secretNS),
+		},
+		Bindings: createBindings(bindings),
+		Template: createListenerTemplate(template),
+	}
+}
+
+// CreateInterceptor returns a GitLab eventlistener
+func (repo *GitLabRepository) CreateInterceptor(secretName, secretNs string) *triggersv1.EventInterceptor {
+	return &triggersv1.EventInterceptor{
+		GitLab: &triggersv1.GitLabInterceptor{
+			SecretRef: &triggersv1.SecretRef{
+				SecretName: secretName,
+				SecretKey:  webhookSecretKey,
+				Namespace:  secretNs,
+			},
+		},
+	}
+}
```

Comment:
There is a lot of room for refactoring with github.go.   I'd create a base type.   Vast majority of the code would wind up in the base type.

Reply:
sure william will look into same and make necessary changes